### PR TITLE
allow to discard unknown entries during indices conversion

### PR DESCRIPTION
**Pull request by zhangxuhong**
- For any null column, we will still pad an entry with the max or last index.
- minor update for TFRecord output, `Array[String]` will be stored as `Array[Array[String]]` with the each of the inner array having size of 1.

### CODE REVIEW

File: avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/TensorizeInJobHelper.scala [hunk 1]
```diff
@@ -230,7 +230,7 @@ object TensorizeInJobHelper {
   private def wrapStringArray: UserDefinedFunction = {
 
     udf {
-      stringSeq: Seq[String] => Seq(stringSeq)
```
Comment (junshi15) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/TensorizeInJobHelper.scala:233]: can you explain why this change is needed?

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala [hunk 2]
```diff
@@ -140,39 +158,48 @@ object FeatureIndicesConversion {
...
    */
-  private def convertNTVToIdValues(ntvs: Seq[Row], featureMapping: Map[String, Long]): Seq[TensorizeIn.IdValue] = {
+  private def convertNTVToIdValues(
+    ntvs: Seq[Row],
+    featureMapping: Map[String, Long],
+    discardUnknownEntries: Boolean): Seq[TensorizeIn.IdValue] = {
 
     // The number of unique name-term combinations
     val cardinality = featureMapping.size.toLong
+    val lastIndex = if (discardUnknownEntries) cardinality - 1 else cardinality
```
Comment (junshi15) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala:171]: please explain this line. Does cardinality include "unknown" or not?
if cardinality counts "unknown" and discardUnknownEntries is true, then lastIndex = cardinality-2.
if cardinality does not count "unknown" and discardUnknownEntries is true, then lastIndex = cardinality-1.
right?

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala [hunk 3]
```diff
@@ -181,18 +208,27 @@ object FeatureIndicesConversion {
...
    */
-  private def convertStringSeqToIdSeq(featureMapping: Map[String, Long]): UserDefinedFunction = {
+  private def convertStringSeqToIdSeq(
+    featureMapping: Map[String, Long],
+    discardUnknownEntries: Boolean): UserDefinedFunction = {
 
     udf {
       stringSeq: Seq[String] => {
-        val cardinality = featureMapping.size.toLong
+        val lastIndex = if (discardUnknownEntries) featureMapping.size.toLong - 1 else featureMapping.size.toLong
```
Comment (junshi15) [inline avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala:220]: does the featureMapping include "unknown" or not?

**Review by junshi15 — COMMENTED** [review]
Please address the issues, otherwise it looks good.

File: avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/TensorizeInJobHelper.scala (hunk 1 above)
Comment (zhangxuhong) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/helpers/TensorizeInJobHelper.scala:233]: There is a bug in the lib to dump TFRecord format. This is a workaround to TFRecord format for string types.

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala (hunk 2 above)
Comment (zhangxuhong) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala:171]: `cardinality` never includes the unknown. so yes, it's cardinality-1 when discarding the unknowns

File: avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala (hunk 3 above)
Comment (zhangxuhong) [reply avro2tf/src/main/scala/com/linkedin/avro2tf/jobs/FeatureIndicesConversion.scala:220]: no feature mapping does not include unknown
